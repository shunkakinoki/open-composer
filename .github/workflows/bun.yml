name: Bun
on:
  push:
    branches:
      - main
  pull_request:
    branches:
      - main
  workflow_dispatch:
concurrency:
  group: ${{ github.workflow }}-${{ github.event_name }}-${{ github.head_ref || github.sha }}
  cancel-in-progress: true
jobs:
  bun-build:
    runs-on: ubuntu-latest
    timeout-minutes: 5
    steps:
      - name: Checkout
        uses: actions/checkout@v5
      - name: Setup Bun
        uses: shunkakinoki/actions/.github/actions/setup-bun@ba7baa1682a0248f78110dff5638e9a70b7703b2
      - name: Build
        run: bun run build
  bun-lint:
    runs-on: ubuntu-latest
    timeout-minutes: 5
    steps:
      - name: Checkout
        uses: actions/checkout@v5
      - name: Setup Bun
        uses: shunkakinoki/actions/.github/actions/setup-bun@ba7baa1682a0248f78110dff5638e9a70b7703b2
      - name: Check
        run: bun run check
      - name: Format
        run: bun run format
      - name: Lint
        run: bun run lint
      - name: Type Check
        run: bun run type-check
  bun-test:
    runs-on: ubuntu-latest
    timeout-minutes: 5
    steps:
      - name: Checkout
        uses: actions/checkout@v5
      - name: Setup Bun
        uses: shunkakinoki/actions/.github/actions/setup-bun@ba7baa1682a0248f78110dff5638e9a70b7703b2
      - name: Test
        run: bun run test
  bun-check:
    if: always()
    needs:
      - bun-build
      - bun-lint
      - bun-test
    runs-on: ubuntu-latest
    timeout-minutes: 3
    steps:
      - name: Alls Green
        uses: re-actors/alls-green@release/v1
        with:
          jobs: ${{ toJSON(needs) }}
