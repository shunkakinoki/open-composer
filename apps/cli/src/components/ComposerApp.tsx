import { Box } from "ink";
import type React from "react";
import { useState } from "react";
import { ChatInterface } from "./ChatInterface.js";
import { CodeEditor } from "./CodeEditor.js";
import { Layout } from "./Layout.js";
import { Sidebar } from "./Sidebar.js";

// Utility function for creating timestamps - can be mocked in tests
export const createTimestamp = () => new Date();
export const getCurrentTime = () => Date.now();

export interface ComposerAppState {
  currentBranch: string;
  currentFile?: string;
  chatMessages: Array<{
    id: string;
    content: string;
    sender: "user" | "agent";
    agent?: string;
    timestamp: Date;
  }>;
  worktrees: Array<{
    name: string;
    path: string;
    branch: string;
    active: boolean;
  }>;
}

export const ComposerApp: React.FC = () => {
  const [state, setState] = useState<ComposerAppState>({
    currentBranch: "main",
    chatMessages: [
      {
        id: "1",
        content: "Welcome to Open Composer CLI! What would you like to build?",
        sender: "agent",
        agent: "composer",
        timestamp: createTimestamp(),
      },
    ],
    worktrees: [
      {
        name: "main",
        path: process.cwd(),
        branch: "main",
        active: true,
      },
    ],
  });

  const handleSendMessage = (message: string) => {
    const newMessage = {
      id: getCurrentTime().toString(),
      content: message,
      sender: "user" as const,
      timestamp: createTimestamp(),
    };

    setState((prev) => ({
      ...prev,
      chatMessages: [...prev.chatMessages, newMessage],
    }));

    // Mock agent response
    setTimeout(() => {
      const agentResponse = {
        id: (getCurrentTime() + 1).toString(),
        content: `I'll help you with: "${message}". Let me analyze your request...`,
        sender: "agent" as const,
        agent: "claude-code",
        timestamp: createTimestamp(),
      };

      setState((prev) => ({
        ...prev,
        chatMessages: [...prev.chatMessages, agentResponse],
      }));
    }, 1000);
  };

  return (
    <Layout>
      <Box flexDirection="row" height="100%">
        <Box width="30%" borderStyle="single" borderColor="gray">
          <Sidebar
            worktrees={state.worktrees}
            currentBranch={state.currentBranch}
          />
        </Box>

        <Box width="40%" borderStyle="single" borderColor="gray">
          <ChatInterface
            messages={state.chatMessages}
            onSendMessage={handleSendMessage}
          />
        </Box>

        <Box width="30%" borderStyle="single" borderColor="gray">
          <CodeEditor currentFile={state.currentFile} />
        </Box>
      </Box>
    </Layout>
  );
};
