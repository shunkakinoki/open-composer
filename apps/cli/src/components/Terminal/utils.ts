/**
 * Terminal utility functions
 * Inspired by termact: https://github.com/MasterGordon/termact/blob/main/src/utils.ts
 */

import type { TerminalStyle } from "./types.js";
import {
  bg,
  blink,
  bold,
  doubleUnderline,
  fg,
  hidden,
  inverse,
  italic,
  strikethrough,
  underline,
} from "./ansi.js";

export const applyStyle = (style: TerminalStyle): string => {
  let result = "";
  if (style.fg) result += fg(style.fg);
  if (style.bg) result += bg(style.bg);
  if (style.bold) result += bold();
  if (style.italic) result += italic();
  if (style.underline) result += underline();
  if (style.blink) result += blink();
  if (style.reverse) result += inverse();
  if (style.hidden) result += hidden();
  if (style.strikethrough) result += strikethrough();
  if (style.doubleUnderline) result += doubleUnderline();
  return result;
};

export const wrapText = (text: string, width: number): string[] => {
  const lines: string[] = [];
  const words = text.split(" ");
  let currentLine = "";

  for (const word of words) {
    if (currentLine.length + word.length + 1 <= width) {
      currentLine += (currentLine ? " " : "") + word;
    } else {
      if (currentLine) lines.push(currentLine);
      currentLine = word;
    }
  }

  if (currentLine) lines.push(currentLine);
  return lines;
};

export const truncateText = (text: string, maxLength: number): string => {
  if (text.length <= maxLength) {
    return text;
  }
  return text.slice(0, maxLength - 3) + "...";
};

export const padText = (
  text: string,
  width: number,
  align: "left" | "center" | "right" = "left",
): string => {
  if (text.length >= width) {
    return text.slice(0, width);
  }

  const padding = width - text.length;

  switch (align) {
    case "center": {
      const leftPad = Math.floor(padding / 2);
      const rightPad = padding - leftPad;
      return " ".repeat(leftPad) + text + " ".repeat(rightPad);
    }
    case "right":
      return " ".repeat(padding) + text;
    default:
      return text + " ".repeat(padding);
  }
};

export const getNtscGrayscale = (r: number, g: number, b: number): number => {
  return r * 0.299 + g * 0.587 + b * 0.114;
};
