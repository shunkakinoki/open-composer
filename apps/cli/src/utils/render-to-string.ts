import { render as inkRender } from "ink";
import type { ReactElement } from "react";
import { EventEmitter } from "node:events";

/**
 * Minimal writable stream for capturing Ink output
 */
class CaptureStream extends EventEmitter {
  private output = "";

  write(chunk: string): void {
    this.output += chunk;
  }

  getOutput(): string {
    return this.output;
  }

  end(): void {
    // no-op
  }
}

/**
 * Renders a React component to a string using Ink
 * Useful for rendering UI components in Effect CLI commands
 */
export function renderToString(element: ReactElement): string {
  const stdout = new CaptureStream();
  const stderr = new CaptureStream();

  const instance = inkRender(element, {
    stdout: stdout as unknown as NodeJS.WriteStream,
    stderr: stderr as unknown as NodeJS.WriteStream,
    debug: false,
    exitOnCtrlC: false,
    patchConsole: false,
  });

  // Wait for a tick to let Ink render
  const output = stdout.getOutput();

  // Clean up
  instance.unmount();
  instance.cleanup();

  return output;
}
