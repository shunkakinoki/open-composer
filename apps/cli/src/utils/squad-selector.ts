import inquirer from "inquirer";
import type { LLMTask } from "@open-composer/agent-registry";
import { getAgentRegistry, getSquadLauncher } from "@open-composer/agent-registry";

/**
 * Start interactive squad selector
 */
export async function startSquadSelector(): Promise<void> {
  const registry = getAgentRegistry();
  const launcher = getSquadLauncher();

  console.log("\n🎮 Pokemon-Style Squad Manager\n");

  const { action } = await inquirer.prompt([
    {
      type: "list",
      name: "action",
      message: "What would you like to do?",
      choices: [
        { name: "📋 List all agents", value: "list" },
        { name: "✨ Create new squad", value: "create" },
        { name: "🚀 Quick squad (by task)", value: "quick" },
        { name: "📊 View statistics", value: "stats" },
        { name: "🚪 Exit", value: "exit" },
      ],
    },
  ]);

  switch (action) {
    case "list": {
      const agents = registry.getAllAgents();
      console.log(`\n📦 Found ${agents.length} agents:\n`);
      for (const agent of agents) {
        console.log(
          `  ${agent.pokemonAttributes.sprite} ${agent.name} (${agent.tier}) - ${agent.provider}`,
        );
      }
      break;
    }

    case "create": {
      console.log("\n🛠️  Create Squad - Not implemented in CLI yet");
      console.log("   Use the interactive UI components for full squad creation");
      break;
    }

    case "quick": {
      const { task } = await inquirer.prompt([
        {
          type: "list",
          name: "task",
          message: "Select task specialization:",
          choices: [
            { name: "🧭 Planner", value: "planner" },
            { name: "⚡ Programmer", value: "programmer" },
            { name: "🔍 Reviewer", value: "reviewer" },
            { name: "🌊 Router", value: "router" },
            { name: "🌿 Summarizer", value: "summarizer" },
          ],
        },
      ]);

      const team = launcher.createQuickSquad(task as LLMTask);
      console.log(`\n✅ Created squad: ${team.name}`);
      console.log(`   Agents: ${team.agents.map((a) => a.name).join(", ")}`);
      break;
    }

    case "stats": {
      const stats = registry.getStats();
      console.log("\n📊 Registry Statistics:");
      console.log(`   Total Agents: ${stats.totalAgents}`);
      console.log(`   Total Squads: ${stats.totalTeams}`);
      console.log(`   Starter: ${stats.tierCounts.starter || 0}`);
      console.log(`   Evolved: ${stats.tierCounts.evolved || 0}`);
      console.log(`   Legendary: ${stats.tierCounts.legendary || 0}`);
      break;
    }

    case "exit":
      console.log("\n👋 Goodbye!\n");
      return;
  }

  // Loop back to main menu
  await startSquadSelector();
}
