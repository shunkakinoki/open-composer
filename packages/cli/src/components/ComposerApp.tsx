// biome-ignore assist/source/organizeImports: False positive
import type React from "react";
import { useState } from "react";
import { Box } from "ink";
import { Layout } from "./Layout.js";
import { ChatInterface } from "./ChatInterface.js";
import { Sidebar } from "./Sidebar.js";
import { CodeEditor } from "./CodeEditor.js";

export interface ComposerAppState {
  currentBranch: string;
  currentFile?: string;
  chatMessages: Array<{
    id: string;
    content: string;
    sender: "user" | "agent";
    agent?: string;
    timestamp: Date;
  }>;
  worktrees: Array<{
    name: string;
    path: string;
    branch: string;
    active: boolean;
  }>;
}

export const ComposerApp: React.FC = () => {
  const [state, setState] = useState<ComposerAppState>({
    currentBranch: "main",
    chatMessages: [
      {
        id: "1",
        content: "Welcome to Open Composer CLI! What would you like to build?",
        sender: "agent",
        agent: "composer",
        timestamp: new Date(),
      },
    ],
    worktrees: [
      {
        name: "main",
        path: process.cwd(),
        branch: "main",
        active: true,
      },
    ],
  });

  const handleSendMessage = (message: string) => {
    const newMessage = {
      id: Date.now().toString(),
      content: message,
      sender: "user" as const,
      timestamp: new Date(),
    };

    setState((prev) => ({
      ...prev,
      chatMessages: [...prev.chatMessages, newMessage],
    }));

    // Mock agent response
    setTimeout(() => {
      const agentResponse = {
        id: (Date.now() + 1).toString(),
        content: `I'll help you with: "${message}". Let me analyze your request...`,
        sender: "agent" as const,
        agent: "claude-code",
        timestamp: new Date(),
      };

      setState((prev) => ({
        ...prev,
        chatMessages: [...prev.chatMessages, agentResponse],
      }));
    }, 1000);
  };

  return (
    <Layout>
      <Box flexDirection="row" height="100%">
        <Box width="30%" borderStyle="single" borderColor="gray">
          <Sidebar
            worktrees={state.worktrees}
            currentBranch={state.currentBranch}
          />
        </Box>

        <Box width="40%" borderStyle="single" borderColor="gray">
          <ChatInterface
            messages={state.chatMessages}
            onSendMessage={handleSendMessage}
          />
        </Box>

        <Box width="30%" borderStyle="single" borderColor="gray">
          <CodeEditor currentFile={state.currentFile} />
        </Box>
      </Box>
    </Layout>
  );
};
