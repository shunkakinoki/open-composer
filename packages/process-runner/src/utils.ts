import * as Effect from "effect/Effect";
import {
  MAX_COMMAND_LENGTH,
  MAX_RUN_NAME_LENGTH,
  RUN_NAME_PATTERN,
} from "./constants.js";
import { ProcessRunnerError } from "./types.js";

// =============================================================================
// VALIDATION UTILITIES
// =============================================================================

export function validateRunName(
  runName: string,
): Effect.Effect<string, ProcessRunnerError> {
  if (!runName || typeof runName !== "string") {
    return Effect.fail(
      ProcessRunnerError("Run name must be a non-empty string"),
    );
  }

  if (runName.length > MAX_RUN_NAME_LENGTH) {
    return Effect.fail(
      ProcessRunnerError(
        `Run name too long (max ${MAX_RUN_NAME_LENGTH} characters)`,
      ),
    );
  }

  if (!RUN_NAME_PATTERN.test(runName)) {
    return Effect.fail(
      ProcessRunnerError(
        "Run name can only contain letters, numbers, hyphens, and underscores",
      ),
    );
  }

  return Effect.succeed(runName);
}

export function validateCommand(
  command: string,
): Effect.Effect<string, ProcessRunnerError> {
  if (!command || typeof command !== "string") {
    return Effect.fail(
      ProcessRunnerError("Command must be a non-empty string"),
    );
  }

  if (command.length > MAX_COMMAND_LENGTH) {
    return Effect.fail(
      ProcessRunnerError(
        `Command too long (max ${MAX_COMMAND_LENGTH} characters)`,
      ),
    );
  }

  // Basic command injection prevention - only reject null bytes
  if (command.includes("\x00")) {
    return Effect.fail(
      ProcessRunnerError("Command contains invalid characters"),
    );
  }

  return Effect.succeed(command);
}

// =============================================================================
// UTILITY FUNCTIONS
// =============================================================================

export function withTimeout<T>(
  effect: Effect.Effect<T, ProcessRunnerError>,
  timeoutMs: number,
  timeoutMessage: string = "Operation timed out",
): Effect.Effect<T, ProcessRunnerError> {
  const timeoutEffect = Effect.async<T, ProcessRunnerError>((resume) => {
    const timeout = setTimeout(() => {
      resume(Effect.fail(ProcessRunnerError(timeoutMessage)));
    }, timeoutMs);
    return Effect.sync(() => clearTimeout(timeout));
  });

  return Effect.race(effect, timeoutEffect);
}
