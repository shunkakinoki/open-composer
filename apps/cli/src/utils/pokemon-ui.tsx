import type { AgentTeam, OpenComposerAgent } from "@open-composer/agent-registry";
import type { RegistryStats } from "@open-composer/analytics";
import { renderToString } from "./render-to-string.js";
import {
  AgentCard,
  AgentComparison,
  AgentList,
  EvolutionTree,
  RegistryStats as RegistryStatsComponent,
  TaskSpecializations,
  TeamDisplay,
} from "../components/PokemonUI.js";

/**
 * Pokemon-style UI renderer using React components
 * This is a utility wrapper that renders React components to strings for CLI output
 */
export class PokemonUI {
  /**
   * Render agent card
   */
  static renderAgentCard(agent: OpenComposerAgent): string {
    return renderToString(<AgentCard agent={agent} />);
  }

  /**
   * Render agent list
   */
  static renderAgentList(agents: OpenComposerAgent[]): string {
    return renderToString(<AgentList agents={agents} />);
  }

  /**
   * Render team display
   */
  static renderTeam(team: AgentTeam): string {
    return renderToString(<TeamDisplay team={team} />);
  }

  /**
   * Render evolution tree
   */
  static renderEvolutionTree(agent: OpenComposerAgent): string {
    return renderToString(<EvolutionTree agent={agent} />);
  }

  /**
   * Render task specializations
   */
  static renderTaskSpecializations(agent: OpenComposerAgent): string {
    return renderToString(<TaskSpecializations agent={agent} />);
  }

  /**
   * Render registry statistics
   */
  static renderStats(stats: RegistryStats): string {
    return renderToString(<RegistryStatsComponent stats={stats} />);
  }

  /**
   * Render agent comparison
   */
  static renderComparison(
    agent1: OpenComposerAgent,
    agent2: OpenComposerAgent,
  ): string {
    return renderToString(<AgentComparison agent1={agent1} agent2={agent2} />);
  }
}
