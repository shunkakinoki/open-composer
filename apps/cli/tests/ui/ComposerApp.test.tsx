import { describe, expect, mock, test } from "bun:test";
import React from "react";

// Mock ComposerApp utility functions before importing
const mockCreateTimestamp = mock(() => new Date("2024-01-01T10:00:00Z"));
const mockGetCurrentTime = mock(() =>
  new Date("2024-01-01T10:00:00Z").getTime(),
);

mock.module("../../src/components/ComposerApp.js", () => ({
  createTimestamp: mockCreateTimestamp,
  getCurrentTime: mockGetCurrentTime,
}));

// Mock CLI version to keep snapshots stable across version changes
mock.module("../../src/lib/version.js", () => ({
  CLI_VERSION: "0.0.0",
}));

import { ComposerApp } from "../../src/components/ComposerApp.js";
import { render } from "../utils.js";

describe("ComposerApp", () => {
  test("renders correctly", async () => {
    const { lastFrame, cleanup } = await render(<ComposerApp />);
    expect(lastFrame()).toMatchSnapshot();
    cleanup();
  });
});
