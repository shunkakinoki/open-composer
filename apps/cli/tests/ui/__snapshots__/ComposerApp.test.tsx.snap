// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`ComposerApp renders correctly 1`] = `
"┌──────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                   🎼 Open Composer CLI v0.0.0                                    │
└──────────────────────────────────────────────────────────────────────────────────────────────────┘
┌────────────────────────────┐┌──────────────────────────────────────┐┌────────────────────────────┐
│                            ││                                      ││                            │
│ 📂 Workspaces              ││ 💬 Chat Interface                    ││ 📝 Code Editor             │
│                            ││                                      ││                            │
│ ● main                     ││ 🎼 composer 10:00                    ││ ┌──────────────────────┐   │
│                            ││  Welcome to Open Composer CLI! What  ││ │                      │   │
│                            ││ would you like to build?             ││ │                      │   │
│ 🌿 Branches                ││                                      ││ │   No file selected   │   │
│                            ││                                      ││ │                      │   │
│ ● main                     ││                                      ││ │                      │   │
│ + New worktree             ││                                      ││ └──────────────────────┘   │
│                            ││                                      ││                            │
│                            ││                                      ││ 🔧 File Tree               │
│ 🤖 Agents                  ││                                      ││                            │
│                            ││                                      ││ 📁 src/                    │
│ ● claude-code              ││ ┌──────────────────────────────────┐ ││   📄 index.ts              │
│ ○ codex-nation             ││ │💭 _                              │ ││   📁 components/           │
│ ○ cursor-agent             ││ └──────────────────────────────────┘ ││     📄 ComposerApp.tsx     │
│                            ││ Type your message and press Enter... ││     📄 ChatInterface.tsx   │
│                            ││                                      ││                            │
└────────────────────────────┘└──────────────────────────────────────┘└────────────────────────────┘
┌──────────────────────────────────────────────────────────────────────────────────────────────────┐
│Ready                                                                               Ctrl+C to exit│
└──────────────────────────────────────────────────────────────────────────────────────────────────┘"
`;
